# Introduction
This notebook will validate that we have correctly set up a Google Colab notebook, that we can create simple dataframes using `pandas`, and that we can create a simple neural network with `tensorflow` and `keras`.

## 1.1 - Setting up the Google Colab Notebook 

We first need to verify that our Colab environment has the required version of Python and Python packages installed:

* Validate that Python is at least 3.6.
* Validate that NumPy is at least 1.18.
* Validate that pandas is at least 1.0.
* Validate that TensorFlow is at least 2.2.

In [1]:
import platform
import numpy as np
import pandas as pd
import tensorflow as tf

print('Python Version: {} '.format(platform.python_version()))
print('NumPy Version: {} '.format(np.__version__))
print('pandas Version: {} '.format(pd.__version__))
print('TensorFlow Version: {} '.format(tf.__version__))

Python Version: 3.6.9 
NumPy Version: 1.19.5 
pandas Version: 1.1.5 
TensorFlow Version: 2.4.1 


## 1.2 - Manipulating Data with pandas



* Create a 4 column DataFrame with 36 rows, the first column being a date field and the rest numbers.

* Fill the first column with the first day of each month for 3 years (for example: 01-01-2018, 02-01-2018).

* Fill the next 2 columns with random numbers.

* Fill the 4th column with the difference of the first 2 data columns (for example: Col3 - Col2).

* Break the DataFrame into 3 different DataFrames based on the dates (for example: 2018, 2019, 2020)

In [3]:
num_cols = 36
df = pd.DataFrame(data = {'Date': pd.date_range('2018-01-01', periods=num_cols, freq='MS'),
                          'Col2': np.random.normal(loc=10, scale=2, size=num_cols),
                          'Col3': np.random.normal(loc=20, scale=1, size=num_cols)})
df['Col4'] = df['Col3'] - df['Col2']

df_2018 = df[df['Date'].dt.year == int(2018)].copy()
df_2019 = df[df['Date'].dt.year == int(2019)].copy()
df_2020 = df[df['Date'].dt.year == int(2020)].copy()

In [4]:
df.info()
df_2018.info()
df_2019.info()
df_2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36 entries, 0 to 35
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Date    36 non-null     datetime64[ns]
 1   Col2    36 non-null     float64       
 2   Col3    36 non-null     float64       
 3   Col4    36 non-null     float64       
dtypes: datetime64[ns](1), float64(3)
memory usage: 1.2 KB
<class 'pandas.core.frame.DataFrame'>
Int64Index: 12 entries, 0 to 11
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Date    12 non-null     datetime64[ns]
 1   Col2    12 non-null     float64       
 2   Col3    12 non-null     float64       
 3   Col4    12 non-null     float64       
dtypes: datetime64[ns](1), float64(3)
memory usage: 480.0 bytes
<class 'pandas.core.frame.DataFrame'>
Int64Index: 12 entries, 12 to 23
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype        

In [5]:
df.head()

Date       Col2       Col3       Col4
0 2018-01-01   9.647392  20.080084  10.432692
1 2018-02-01   7.312647  20.277915  12.965267
2 2018-03-01  12.827084  19.423024   6.595939
3 2018-04-01  10.067780  20.008374   9.940593
4 2018-05-01   6.050964  19.594223  13.543259

In [6]:
df_2018.head()

Date       Col2       Col3       Col4
0 2018-01-01   9.647392  20.080084  10.432692
1 2018-02-01   7.312647  20.277915  12.965267
2 2018-03-01  12.827084  19.423024   6.595939
3 2018-04-01  10.067780  20.008374   9.940593
4 2018-05-01   6.050964  19.594223  13.543259

In [9]:
df_2019.head()

Date       Col2       Col3       Col4
12 2019-01-01   9.652321  21.695925  12.043604
13 2019-02-01   7.773237  21.919813  14.146576
14 2019-03-01   8.679719  20.790891  12.111172
15 2019-04-01   5.917910  21.804461  15.886551
16 2019-05-01  12.415210  20.068767   7.653557

In [10]:
df_2020.head()

Date       Col2       Col3      Col4
24 2020-01-01   8.518070  18.266322  9.748252
25 2020-02-01  10.956408  19.427713  8.471305
26 2020-03-01  10.654549  19.680034  9.025485
27 2020-04-01  11.043450  19.488194  8.444743
28 2020-05-01  10.880598  20.649653  9.769055

## Create a Neural Network that can Convert Celsius to Fahrenheit

Basic understanding of TensorFlow and Keras will be demonstrated as a prerequisate for this Manning liveProject.

* Using TensorFlow and Keras build a single layered network with a single input and a single output and a single neuron.
* Using NumPy generate ~20 input/output values to test
* Train the neural network on your data.
* Test the neural network to see if it is working.
* Examine the weights to see if they match F = C * 1.8 + 32F=C∗1.8+32.

In [ ]:
from tensorflow import keras

celsius = np.array([x*10 for x in range (-10, 15)])
fahrenheit = (celsius * 1.8) + 32

l0 = tf.keras.layers.Dense(units=1, input_shape=[1])
model = tf.keras.Sequential([l0])

model.compile(loss='mean_squared_error',
              optimizer=tf.keras.optimizers.Adam(0.1))

model.fit(celsius, fahrenheit, epochs=500, verbose=False)
print(model.predict([100.0]))

[[209.37314]]


In [ ]:
print("Layer Weight (W): {}".format(l0.get_weights()[0]))
print("Layer Bias (b): {}".format(l0.get_weights()[1]))

Layer Weight (W): [[1.8148566]]
Layer Bias (b): [27.887472]


From above, our equation for converting fahrenheit to celsius is F = C * 1.815 + 27.887F.  Not bad for the small amount of data used to train our network!